SPUserManager: fix crash with MariaDB 10.3

noahwilliamsson opened the pull request:
Patch by @Muffinman in #3167 for the MariaDB 10.3.5 change at https://github.com/MariaDB/server/commit/f7621f17bd (Truncate_versioning_priv->Delete_history_priv).

Closes #2982, closes #3139, closes #3133 and closes #3167.

dmoagx commented:
The interesting part here is not the SPUserModel definition (if it were just that I would already have fixed it back here https://github.com/sequelpro/sequelpro/issues/2982#issuecomment-365021771 ), but all the consequences this field may have to all other queries of the user manager, e.g. 
* does adding users still work?
* does changing users still work?
* will the value of the field be preserved through any unrelated changes? 
* are there other queries and/or different MySQL versions this may have a side effect on?

noahwilliamsson commented:
@dmoagx Thank's for the feedback.  Great questions!  I'm sorry I did not think of addressing these things in the PR.

- Yes, adding users still work as expected.  Tested with MariaDB 10.2.15 and MariaDB 10.3.9.  The `Delete_history_field` column defaults to `N` and is not touched by Sequel Pro.  I verified this using both `mysqldump` and by screening Sequel Pro's console log.
- Yes, changing users still work as expected.  For each of MariaDB 10.2 and 10.3, I modified users where the field was both `N` and `Y`.  Sequel Pro did not touch this field.
- Yes, it appears as if Sequel Pro does not touch it.  I modified most checkboxes and fields inside the User manager without seeing Sequel Pro getting near that specific field.

It is my understanding that this field will not be touched because it's currently not understood by Sequel Pro.  For this reason I don't think this patch will have any side effects on other database engines.

This patch will merely prevent Sequel Pro from crashing but not properly solve the underlying issue.


After taking a closer look at `SPUserManager.m`,  it seems that `privColumnToGrantMap` might need an update to actually make Sequel Pro understand the field:

- https://github.com/sequelpro/sequelpro/blob/master/Source/SPUserManager.m#L95

Also: `SHOW PRIVILEGES` returns [`Privilege:Delete versioning rows`, `Context:Tables`, `Comment:To delete versioning table historical rows`] in MariaDB 10.3.9.

However, after a naive `@"Delete_history_priv": @"Delete_versioning_rows",` addition to `privColumnToGrantMap` I get a crash in `<SPPrivilegesMO 0x...> setValue:forUndefinedKey:] the entity Privileges is not key value coding-comp`.

I'm assuming I also need to modify `Interfaces/English.lproj/UserManagerView.xib`.  Unfortunately IB is out of my comfort zone. :)

I would be happy to provide access to a MariaDB 10.3 server if that would help speed up a proper solution.

dmoagx commented:
@noahwilliamsson Thanks for looking into it.

I think the xib would only need changes if we wanted to actually add a checkbox for the setting (which would mean we would also have to dynamically hide it if not supported by the server version).

I'll merge this PR and look into the other stuff.

dmoagx commented:
As always it turns out to be much more complicated:

MariaDB first added column `Delete_versioning_rows_priv` with GRANT privilege `DELETE VERSIONING ROWS` but never released that.
Then they renamed it to `Truncate_versioning_priv` and GRANT privilege `DELETE HISTORY` and released it with 10.3.4.
Then they renamed it AGAIN, but this time only the column to `Delete_history_priv` and released it with 10.3.5.

To add even more confusion, their manual suggests that both `DELETE HISTORY` and `DELETE VERSIONING ROWS` are valid names for the GRANT privilege (but `DELETE HISTORY` is preferred), while `SHOW PRIVILEGES` only returns `Delete versioning rows`.

Apart from the whole situation being rather silly, that also means this PR is actually wrong, because it uses `delete_history_priv` but it should actually use `delete_versioning_rows_priv`.
But even then I'm not sure what would work on 10.3.4 and what on 10.3.5+

Muffinman commented:
Oh gosh what a mess!

Perhaps the correct approach is just to 'handle' the case where sequelpro does not recognise the name of a particular permission, rather then attempting to work around every possible variation of the permission name on every version of MariaDB?

noahwilliamsson commented:
@dmoagx MariaDB 10.3 became a GA (generally available; stable) release[1] with 10.3.7 in late May, 2018 so  I wouldn't worry about anything before that.

Oops, that sucks.  I thought I carefully checked this exact thing against the `Create_tmp_table_priv` yesterday but it looks like you're right.

[1] https://mariadb.org/mariadb-10-3-7-now-available/

dmoagx commented:
> Perhaps the correct approach is just to 'handle' the case where sequelpro does not recognise the name of a particular permission

Yes, but more work -> more time ;)

Muffinman commented:
BTW I just checked a new install of 10.3.9 and that permission is still named `Delete_history_priv`?

https://github.com/MariaDB/server/commit/f4270fc544
https://github.com/MariaDB/server/commit/f7621f17bd8a3dd08fae6ad023716cc66412e1b1

![screen shot 2018-10-02 at 22 55 43](https://user-images.githubusercontent.com/1319568/46379485-c3f7ef80-c696-11e8-9376-d0ea5cc3d70a.png)

dmoagx commented:
No, that's what I mentioned above.
The underlying column is named `Delete_history_prev`, but the privilege returned by "SHOW PRIVILEGES" is called `Delete versioning rows`.
Sequel Pro needs to match up both (well, at least if we were to actually display it in the GUI).
